### PR TITLE
feat: collect linker metadata if available

### PR DESCRIPTION
**Pull request by eseidel**
Runs `aot_tools link_metadata` if available.  Does not block the build if it is not.

### CODE REVIEW

File: packages/shorebird_cli/lib/src/platform/apple.dart [hunk 1]
```diff
@@ -26,7 +27,30 @@ enum ApplePlatform {
...
+
+  /// Creates a new [LinkResult] representing success.
+  const LinkResult.success({required this.linkPercentage, this.linkMetadata})
+    : _exitCodeObject = ExitCode.success;
+
+  /// ExitCode.code isn't const, so store the actual object.
+  final ExitCode _exitCodeObject;
+
+  /// The exit code of the linker process.
+  int get exitCode => _exitCodeObject.code;
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/platform/apple.dart:46]: why do we need to hold onto the object instead of just the underlying int?

File: packages/shorebird_cli/lib/src/platform/apple.dart [hunk 2]
```diff
@@ -26,7 +27,30 @@ enum ApplePlatform {
 
 /// A record containing the exit code and optionally link percentage
 /// returned by `runLinker`.
-typedef LinkResult = ({int exitCode, double? linkPercentage});
+@immutable
+class LinkResult {
+  /// Creates a new [LinkResult] representing failure.
+  const LinkResult.failure()
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/platform/apple.dart:33]: wouldn't we want to have an error reason (e.g. from `stderr`) in the case of a failure?

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 1 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:46]: We don't.  The only advantage of this is to let us have const constructors.

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 2 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:33]: We certainly could.  I was just matching what the existing code did, just moving it from being a record type to an explicit class with a named constructor (with appropriate defaults).

File: packages/shorebird_cli/lib/src/platform/apple.dart [hunk 3]
```diff
@@ -289,12 +313,30 @@ $error''');
       );
     } on Exception catch (error) {
       linkProgress.fail('Failed to link AOT files: $error');
-      return (exitCode: ExitCode.software.code, linkPercentage: null);
+      return const LinkResult.failure();
     } finally {
       await dumpDebugInfo();
     }
+    Map<String, dynamic>? linkMetadata;
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/platform/apple.dart:320]:
```suggestion
    final Map<String, dynamic>? linkMetadata;
```

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 1 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:46]: The reason why we can't have a const constructor if we hold onto the int directly is that ExitCode.code isn't const-compatible.

File: packages/shorebird_cli/test/src/executables/aot_tools_test.dart [hunk 4]
```diff
@@ -950,6 +950,69 @@ Run "aot_tools help <command>" for more information about a command.
...
+            ).thenAnswer((_) => Stream.value(utf8.encode(stderr)));
+            return mockProcess;
+          });
+          when(
+            () => process.start(
+              any(),
+              any(),
+              workingDirectory: any(named: 'workingDirectory'),
+            ),
+          ).thenAnswer((invocation) async {
```
Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart]:
```suggestion
          ).thenAnswer((_) async {
```

File: packages/shorebird_cli/test/src/executables/aot_tools_test.dart [hunk 5]
```diff
@@ -950,6 +950,69 @@ Run "aot_tools help <command>" for more information about a command.
...
+            ).thenAnswer((_) => Stream.value(utf8.encode(stdout)));
+            when(
+              () => mockProcess.stderr,
+            ).thenAnswer((_) => Stream.value(utf8.encode(stderr)));
+            return mockProcess;
+          });
+        });
+
+        test('returns link metadata', () async {
+          stdout = '{}';
```
Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart:999]: why is this stdout assignment needed?

File: packages/shorebird_cli/test/src/executables/aot_tools_test.dart [hunk 6]
```diff
@@ -950,6 +950,69 @@ Run "aot_tools help <command>" for more information about a command.
           expect(result.existsSync(), isTrue);
         });
       });
+
+      group('getLinkMetadata', () {
+        late int exitCode;
+        late String stdout;
+        late String stderr;
+        setUp(() {
```
Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart]:
```suggestion
        late String stderr;

        setUp(() {
```
nit

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 3 above)
Comment (eseidel) [inline packages/shorebird_cli/lib/src/platform/apple.dart:320]: We can't as I've currently written it, since the analyzer can't be sure it' isn't set twice.

File: packages/shorebird_cli/test/src/executables/aot_tools_test.dart [hunk 7]
```diff
@@ -950,6 +950,69 @@ Run "aot_tools help <command>" for more information about a command.
...
+              () => mockProcess.stdout,
+            ).thenAnswer((_) => Stream.value(utf8.encode(stdout)));
+            when(
+              () => mockProcess.stderr,
+            ).thenAnswer((_) => Stream.value(utf8.encode(stderr)));
+            return mockProcess;
+          });
+        });
+
+        test('returns link metadata', () async {
```
Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart]:
```suggestion
        test('returns link metadata when aot_tools executes successfully', () async {
```

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 3 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:320]: the `if (dumpDebugInfoDir == null)` branch seems to confuse it into believing it might be set twice.

File: packages/shorebird_cli/test/src/executables/aot_tools_test.dart [hunk 8]
```diff
@@ -950,6 +950,69 @@ Run "aot_tools help <command>" for more information about a command.
...
+          stdout = '{}';
+          final result = await runWithOverrides(
+            () => aotTools.getLinkMetadata(debugDir: '/debug'),
+          );
+          expect(result, isA<Map<String, dynamic>>());
+        });
+
+        test('returns invalid json', () async {
```
Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart]:
```suggestion
        test('throws FormatException when aot_tools outputs invalid json', () async {
```

Comment (felangel) [inline packages/shorebird_cli/test/src/executables/aot_tools_test.dart]: Shouldn't we also have a test to ensure that when aot_tools exits with a non-zero exit code we treat it as a failure?

Comment (eseidel) [reply packages/shorebird_cli/test/src/executables/aot_tools_test.dart]: We don't treat it as a failure.  If we can't get the metadata we just keep going.  It doesn't block the user.

File: packages/shorebird_cli/lib/src/platform/apple.dart [hunk 9]
```diff
@@ -289,12 +313,30 @@ $error''');
...
+    try {
+      if (dumpDebugInfoDir == null) {
+        linkMetadata = null;
+      } else {
+        linkMetadata = await aotTools.getLinkMetadata(
+          debugDir: dumpDebugInfoDir.path,
+          workingDirectory: buildDirectory.path,
+        );
+      }
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/platform/apple.dart]:
```suggestion
      if (dumpDebugInfoDir != null) {
        linkMetadata = await aotTools.getLinkMetadata(
          debugDir: dumpDebugInfoDir.path,
          workingDirectory: buildDirectory.path,
        );
      }
```

File: packages/shorebird_cli/lib/src/platform/apple.dart [hunk 10]
```diff
@@ -289,12 +313,30 @@ $error''');
...
+    } on Exception catch (error) {
+      logger.detail('Failed to get link metadata: $error');
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/platform/apple.dart]:
```suggestion
      logger.detail('[aot_tools] Failed to get link metadata: $error');
```
nit

File: packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart [hunk 11]
```diff
@@ -175,13 +179,22 @@ class IosFrameworkPatcher extends Patcher {
...
         vmCodeFile: File(_vmcodeOutputPath),
       );
+      final linkPercentage = result.linkPercentage;
+      final exitCode = result.exitCode;
+      if (exitCode != ExitCode.success.code) throw ProcessExit(exitCode);
+      if (linkPercentage != null &&
+          linkPercentage < Patcher.linkPercentageWarningThreshold) {
+        logger.warn(Patcher.lowLinkPercentageWarning(linkPercentage));
+      }
+      lastBuildLinkPercentage = linkPercentage;
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: nice so this was never being reported before?

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 2 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:33]: yeah makes sense, you can ignore my comment 👍

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 1 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:46]: Isn't the failure case's exit code always software? Can't we avoid passing an ExitCode altogether?

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 3 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:320]: yeah makes sense it's fine as is 👍

File: packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart (hunk 11 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: Correct, the ios_framework side was not reporting link percentages.

Comment (eseidel) [reply packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: Mostly all of this just needs to be pulled out of these leaf classes and into some shared logic.

Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: probably should add a unit test to make sure there's no regression in the future 🤷

File: packages/shorebird_cli/lib/src/platform/apple.dart (hunk 1 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/platform/apple.dart:46]: Do you mean just store a bool instead of the code?

File: packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart (hunk 11 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: Happy to.

Comment (eseidel) [reply packages/shorebird_cli/lib/src/commands/patch/ios_framework_patcher.dart:196]: Done.
